File: backend/myPizza/src/main/java/backendapp/myPizza/services/ProductService.java
package backendapp.myPizza.services;

import backendapp.myPizza.Models.entities.Product;
import backendapp.myPizza.Models.entities.Topping;
import backendapp.myPizza.Models.reqDTO.ManyProductsPostDTO;
import backendapp.myPizza.Models.reqDTO.ProductDTO;
import backendapp.myPizza.Models.reqDTO.ToppingDTO;
import backendapp.myPizza.Models.resDTO.ConfirmRes;
import backendapp.myPizza.exceptions.BadRequestException;
import backendapp.myPizza.repositories.ProductRepository;
import backendapp.myPizza.repositories.ToppingRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;


@Service
public class ProductService {

    @Autowired
    private ToppingRepository toppingRp;

    @Autowired
    private ProductRepository productRp;

    private static final DecimalFormat df = new DecimalFormat("0.00");

    public List<Topping> findAllToppings() {
        return toppingRp.findAll().stream()
                .peek(t -> t.setDescription(t.getName() + " (" + df.format(t.getPrice()) + "€)")).toList();
    }

    public Topping addTopping(ToppingDTO t) throws BadRequestException {
        List<String> existingToppingNames = toppingRp.findAllToppingNames();
        // Duplicate exception handling
        if (existingToppingNames.contains(t.name()))
            throw new BadRequestException("Topping with name '" + t.name() + "' already exists");
        Topping topping = new Topping(t.name(), t.price());
        toppingRp.save(topping);
        topping.setDescription(topping.getName() + " (" + df.format(topping.getPrice()) + "€)");
        return topping;
    }

    public Topping updateToppingByName(String oldName, ToppingDTO toppingDTO) throws BadRequestException {
        Topping oldTopping = toppingRp.findByName(oldName).orElseThrow(
                () -> new BadRequestException("Topping you're trying to update doesn't exist")
        );
        toppingRp.delete(oldTopping);
        Topping topping = new Topping(toppingDTO.name(), toppingDTO.price());
        toppingRp.save(topping);
        topping.setDescription(topping.getName() + " (" + df.format(topping.getPrice()) + "€)");
        return topping;
    }


    public ConfirmRes deleteToppingByName(String name) throws BadRequestException {
        Topping topping = toppingRp.findByName(name).orElseThrow(
                () -> new BadRequestException("Topping you're trying to delete doesn't exist")
        );
        for (Product p : topping.getProducts()) {
            p.getToppings().remove(topping);
            productRp.save(p);
        }
        toppingRp.delete(topping);
        return new ConfirmRes("Topping '" + name + "' has been successfully deleted", HttpStatus.OK);
    }

    public Page<Product> getAllProducts(Pageable pageable) {
        return productRp.findAll(pageable).map(p -> {
            p.setToppings(p.getToppings().stream()
                    .peek(t -> t.setDescription(t.getName() + " (" + df.format(t.getPrice()) + "€)")).toList());
            p.setProductTotalAmount();
            return p;
        });
    }

    public Product updateProductByName(String oldName, ProductDTO productDTO) throws BadRequestException {

        Product oldProduct = productRp.findByName(oldName).orElseThrow(
                () -> new BadRequestException("The product you're trying to update (name: '" + oldName + "') doesn't exist")
        );
        productRp.delete(oldProduct);

        Product product = new Product(productDTO.name(), productDTO.basePrice(), productDTO.category());

        List<Topping> toppings = new ArrayList<>();

        for (String name : productDTO.toppings()) {
            Topping t = toppingRp.findByName(name).orElseThrow(
                    () -> new BadRequestException("Topping (old name: '" + name + "') you're trying to update doesn't exist")
            );
            t.setDescription(t.getName() + " (" + df.format(t.getPrice()) + "€)");
            toppings.add(t);
        }

        for (Topping t : toppings) {
            product.getToppings().add(t);
        }

        product.setProductTotalAmount();
        productRp.save(product);
        return product;
    }

    public ConfirmRes deleteProductByName(String name) throws BadRequestException {
        Product product = productRp.findByName(name).orElseThrow(
                () -> new BadRequestException("The product you're trying to delete doesn't exist")
        );
        productRp.delete(product);
        return new ConfirmRes("Product with name '" + name + "' deleted successfully", HttpStatus.OK);
    }

    public List<String> getAllCategories() {
        List<String> categories = productRp.getAllCategories();
        categories.addFirst("(seleziona una categoria)");
        categories.addLast("- Inserisci nuova -");
        return categories;
    }

    public List<String> getProductNames() {
        return productRp.findAll().stream().map(Product::getName).toList();
    }

    public List<Product> addManyProducts(ManyProductsPostDTO manyProductsPostDTO) throws BadRequestException {
        List<ProductDTO> products = manyProductsPostDTO.products();
        // Handling exceptions before making db transactions
        for (ProductDTO p : products) {
            Optional<Product> product = productRp.findByName(p.name());
            if (product.isPresent())
                throw new BadRequestException("Product with name '" + p.name() + "' already exists");
            List<String> toppingsNames = p.toppings();
            for (String name : toppingsNames) {
                toppingRp.findByName(name).orElseThrow(
                        () -> new BadRequestException("Topping '" + name + "' doesn't exist")
                );
            }
        }
        // Handling exception finish
        List<Product> addedProducts = new ArrayList<>();
        for (ProductDTO p : products) {
            Product newProduct = new Product(p.name(), p.basePrice(), p.category());


            for (String name : p.toppings()) {
                assert toppingRp.findByName(name).isPresent();
                newProduct.getToppings().add(toppingRp.findByName(name).get());
            }
            addedProducts.add(newProduct);
        }
        productRp.saveAll(addedProducts);
        return addedProducts.stream()
                .peek(p -> p.setToppings(p.getToppings().stream()
                        .peek(t -> t.setDescription(t.getName() + " (" + df.format(t.getPrice()) + "€)"))
                        .toList())).toList();
    }

    public Product addProduct(ProductDTO productDTO) throws BadRequestException {
        Product p = new Product();
        p.setName(productDTO.name());
        p.setBasePrice(productDTO.basePrice());
        for (String toppingName : productDTO.toppings()) {
            Topping topping = toppingRp.findByName(toppingName)
                    .orElseThrow(
                            () -> new BadRequestException("Topping with name '" + toppingName + "doesn't exist")
                    );
        }
        for (String toppingName : productDTO.toppings()) {
            assert toppingRp.findByName(toppingName).isPresent();
            Topping topping = toppingRp.findByName(toppingName).get();
            p.getToppings().add(topping);
        }
        p.setCategory(productDTO.category());
        p.setProductTotalAmount();
        p.setToppings(p.getToppings().stream()
                .peek(t -> t.setDescription(t.getName() + " (" + df.format(t.getPrice()) + "€)")).toList());

        return productRp.save(p);

    }

}

File: backend_nodejs_nest_fastify/my-pizza/src/nest_modules/product/services/product.service.ts
import { BadRequestException, HttpStatus, Injectable, InternalServerErrorException, Logger } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Topping } from '../entities/topping.entity';
import { DataSource, Repository } from 'typeorm';
import { Product } from '../entities/product.entity';
import { Category } from '../entities/category.entity';
import { Menu } from '../entities/menu.entity';
import { ToppingRes } from '../interfaces/topping-res.interface';
import { ToppingType } from '../enums/topping-type.enum';
import { ToppingDTO } from '../interfaces/topping-dto.interface';
import { ConfirmRes } from 'src/nest_modules/auth-user/interfaces/confirm-res.interface';
import { paginate, Pagination, IPaginationOptions } from 'nestjs-typeorm-paginate';
import { ManyproductsPostDTO } from '../interfaces/many-products-post-dto.interface';
import { ProductRes } from '../interfaces/product-res.interface';
import { ProductDTO } from '../interfaces/product-dto.interface';


@Injectable()
export class ProductService {

    private logger = new Logger('ProductService')

    constructor(
        @InjectRepository(Topping) private toppingRepository: Repository<Topping>,
        @InjectRepository(Product) private productRepository: Repository<Product>,
        @InjectRepository(Category) private categoryRepository: Repository<Category>,
        @InjectRepository(Menu) private menuRepository: Repository<Menu>,
        private dataSource: DataSource
    ) { }

    // ._._._._._._._._._._._._._._ TOPPINGS ._._._._._._._._._._._._._._

    public async findAllToppings(type?: ToppingType): Promise<ToppingRes[]> {
        return (await this.toppingRepository.find()).filter(topping => {
            if (type) {
                return topping.type === type
            }
            return true
        }).map(topping => this.generateToppingResModel(topping))
    }

    // private async findAllToppingNames(): Promise<string[]> {
    //     return (await this.toppingRepository
    //         .createQueryBuilder('topping')
    //         .select(['topping.name'])
    //         .getMany()).map(topping => topping.name)
    // }

    private generateToppingResModel(topping: Topping): ToppingRes {
        topping.setDescription()
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const { setDescription, products, id, ...toppingRes } = topping
        return { id, ...toppingRes }
    }

    private async findToppingByName(name: string): Promise<Topping | null | undefined> {
        return await this.toppingRepository.findOneBy({ name })
    }

    public async addTopping(toppingDTO: ToppingDTO): Promise<ToppingRes> {

        const topping = new Topping(toppingDTO.name, toppingDTO.price, toppingDTO.type)

        try {
            await this.toppingRepository.save(topping)
        } catch (e) {
            if (e.message) {
                if (typeof e.message === 'string') {
                    if (e.message.includes('Duplicate entry'))
                        throw new BadRequestException(`topping with name=${toppingDTO.name} already exist`)
                }
            }
            throw new InternalServerErrorException('Database error')
        }
        return this.generateToppingResModel(topping)
    }

    public async updateToppingByName(oldName: string, toppingDTO: ToppingDTO): Promise<ToppingRes> {
        const topping: Topping | null | undefined = await this.findToppingByName(oldName)
        if (!topping) throw new BadRequestException(`topping with name=${oldName} doesn't exist`,
            { cause: new Error(), description: 'Bad Request' })
        topping.name = toppingDTO.name
        topping.price = toppingDTO.price
        topping.type = toppingDTO.type
        try {
            await this.toppingRepository.save(topping)
        } catch (e) {
            if (e.message) {
                if (typeof e.message === 'string') {
                    if (e.message.includes('Duplicate entry'))
                        throw new BadRequestException(`topping with name=${toppingDTO.name} already exist`)
                }
            }
            throw new InternalServerErrorException('Database error')
        }
        topping.setDescription()
        return this.generateToppingResModel(topping)
    }

    public async deleteToppingByName(name: string): Promise<ConfirmRes> {

        const topping: Topping | null | undefined = await this.findToppingByName(name)

        if (!topping) throw new BadRequestException(`Topping with name='${name}' you're trying to delete doesn't exist`)

        if (topping.products) {
            for await (const product of topping.products) {
                const i: number = product.toppings?.indexOf(topping)
                if (i !== - 1) product.toppings?.splice(i, 1)
                await this.productRepository.save(product)
            }
        }

        await this.toppingRepository.delete(topping.id)

        return {
            statusCode: HttpStatus.OK,
            timestamp: new Date().getTime(),
            message: `topping with name='${name}' has been successfully deleted`
        }

    }

    // ._._._._._._._._._._._._._._ CATEGORIES ._._._._._._._._._._._._._._

    public generateCategoryResModel(category: Category): Category {
        const {id, ...otherProps} = category
        return {
            id,
            ...otherProps
        }
    }
    
    private async findCategoryByName(name: string): Promise<Category | null | undefined> {
        return await this.categoryRepository.findOneBy({ name })
    }

    public async isPresentCategory(name: string): Promise<boolean> {
        return !!await this.findCategoryByName(name)
    }

    public async getAllCategoryNames(): Promise<string[]> {
        const results = (await this.categoryRepository
            .createQueryBuilder('category')
            .select(['category.name'])
            .getMany()).map(category => category.name)
        results.unshift('(seleziona una categoria)')
        results.push('- Inserisci nuova -')
        return results
    }

    public async productsHaveNotCategory(category: Category): Promise<boolean> {
        return !await this.productRepository
            .createQueryBuilder('product')
            .where('product.category = :category_id', { category_id: category.id })
            .getCount()
    }

    // ._._._._._._._._._._._._._._ PRODUCTS ._._._._._._._._._._._._._._

    public generateProductResModel(product: Product): ProductRes {
        
        product.setProductTotalAmount()
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const { setProductTotalAmount, id, toppings, category, ...productResOthers } = product
        const toppingResList: ToppingRes[] = []
        toppings.forEach(topping => toppingResList.push(this.generateToppingResModel(topping)))
        
        return {
            id, // per farlo apparire per primo
            ...productResOthers,
            category: this.generateCategoryResModel(category),
            toppings: toppingResList
        }

    }

    public async getAllProducts(options: IPaginationOptions): Promise<Pagination<ProductRes>> {
        const results = await paginate<Product>(this.productRepository, options)
        return new Pagination<ProductRes>(
            results.items.map(product => this.generateProductResModel(product)),
            results.meta
        )
    }

    public async findProductByName(name: string): Promise<Product | null | undefined> {
        return this.productRepository.findOneBy({ name })
    }

    public async addManyProducts(manyProductsPostDTO: ManyproductsPostDTO): Promise<ProductRes[]> {

        // Exception handling start
        for await (const productDTO of manyProductsPostDTO.products) {
            if (!!await this.findProductByName(productDTO.name))
                throw new BadRequestException(`product with name=${productDTO.name} already exist`,
                    { cause: new Error(), description: 'Bad Request' })

            for await (const name of productDTO.toppings) {
                if (!await this.toppingRepository.findOneBy({ name }))
                    throw new BadRequestException(`topping with name=${name} doesn't exist`,
                        { cause: new Error(), description: 'Bad Request' })
            }
        }
        // Exception handling finish

        const addedProducts: Product[] = []

        for await (const productDTO of manyProductsPostDTO.products) {

            let category: Category

            if (await this.isPresentCategory(productDTO.category)) {
                category = await this.findCategoryByName(productDTO.category)
            } else {
                category = new Category(productDTO.category)
                await this.categoryRepository.save(category)
                await this.menuRepository.save(new Menu(category))
            }

            const newProduct = new Product(productDTO.name, productDTO.basePrice, category)

            newProduct.toppings = []
            /* Altrimenti è undefined se non ci sono topping, 
            questa è una differenza sostanziale di TypeOrm rispetto ad Hibernate che invece dà una lista vuota */

            for await (const name of productDTO.toppings) {
                newProduct.toppings.push(await this.findToppingByName(name))
            }

            addedProducts.push(newProduct)
        }

        try {
            await this.productRepository.save(addedProducts)
        } catch {
            throw new InternalServerErrorException('Database error')
        }

        await this.productRepository.save(addedProducts)

        const menuProducts: Menu[] = addedProducts.map(product => new Menu(product))

        await this.menuRepository.save(menuProducts)

        return addedProducts.map(product => this.generateProductResModel(product))

    }


    public async updateProductByName(oldName: string, productDTO: ProductDTO): Promise<ProductRes> {

        const product = await this.findProductByName(oldName)

        if (!product)
         throw new BadRequestException(`the product with name='${oldName}' you're trying to update doesn't exist`)
           
        const menu: Menu = await this.menuRepository.findOneBy({ item: { id: product.id } })
        await this.menuRepository.delete(menu.id)

        const sentCategory: Category | null | undefined = await this.findCategoryByName(productDTO.category)
        const oldCategory: Category = product.category


        if (sentCategory) {
            if (sentCategory !== product.category) {
                product.category = sentCategory
                await this.productRepository.save(product)

            }
        } else {
            const newCategory = new Category(productDTO.category)
            await this.categoryRepository.save(newCategory)
            await this.menuRepository.save(new Menu(newCategory))
            product.category = newCategory
            await this.productRepository.save(product)

        }

        if (await this.productsHaveNotCategory(oldCategory)) {
            await this.categoryRepository.delete(oldCategory.id)
            await this.menuRepository.delete({ item: { id: oldCategory.id } })
        }

        product.name = productDTO.name
        product.basePrice = productDTO.basePrice

        product.toppings = []

        for await (const name of productDTO.toppings) {
            const topping: Topping | null | undefined = await this.findToppingByName(name)
            if (!topping)
                throw new BadRequestException(`topping with name='${name}' you're trying to update doesn't exist`, {
                    cause: new Error(), description: 'Bad request'
                })
            product.toppings.push(topping)
        }


        try {
            await this.productRepository.save(product)
        } catch (e) {
            if (e.message) {
                if (typeof e.message === 'string') {
                    if (e.message.includes('Duplicate entry'))
                        throw new BadRequestException(`product with name='${product.name}' already exist`)
                }
            }
            throw new InternalServerErrorException('Database error')
        } finally {
            await this.menuRepository.save(new Menu(product))
        }

        return this.generateProductResModel(product)

    }


    public async getProductNames(): Promise<string[]> {
        return (await this.productRepository
            .createQueryBuilder('product')
            .select(['product.name'])
            .getMany()).map(product => product.name)
    }



    public async deleteProductByName(name: string): Promise<ConfirmRes> {
        const product: Product | null | undefined = await this.productRepository.findOneBy({ name })
        if (!product) throw new BadRequestException(`product with name='${name}' you're trying to delete doesn't exist`)
        const category: Category = product.category
        const menu = await this.menuRepository.findOneBy({ item: { id: product.id } })
        await this.menuRepository.delete(menu.id)
        await this.productRepository.delete(product.id)
        if (await this.productsHaveNotCategory(category)) {
            const menu = await this.menuRepository.findOneBy({ item: { id: category.id } })
            await this.menuRepository.delete(menu.id)
            await this.categoryRepository.delete(category.id)
        }

        return {
            statusCode: HttpStatus.OK,
            timestamp: new Date().getTime(),
            message: `product with name='${name}' has been deleted successfully`
        }

    }



}
